### PR TITLE
[build]: Adding support for Free-Range-Routing stack.

### PR DESCRIPTION
**Pull request by rodnymolina**
- Extending SONiC building infrastructure to provide users
           with greater flexibility, by allowing them to elect a
           routing-stack different than the default one (quagga). The desired
           routing-stack will be defined in rules/config file.

         - As part of these changes I'm adding support for
           Free-Range-Routing (FRR) stack. Quagga will continue to be
           the default routing-stack.

Signed-off-by: Rodny Molina <rodny@linkedin.com>

### CODE REVIEW

File: dockers/docker-fpm-frr/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+RUN rm -rf /debs
+
+COPY ["*.j2", "/usr/share/sonic/templates/"]
+COPY ["start.sh", "config.sh", "/usr/bin/"]
+COPY ["daemons", "/etc/frr/"]
+COPY ["debian.conf", "/etc/frr/"]
+
+ENTRYPOINT /usr/bin/config.sh   \
+        && /usr/bin/start.sh    \
+        && /bin/bash
```
Comment (stcheng) [inline dockers/docker-fpm-frr/Dockerfile]: Dockerfile.j2 is already in the commit. The generated Dockerfile is not needed.

File: dockers/docker-fpm-quagga/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+RUN apt-get clean -y; apt-get autoclean -y; apt-get autoremove -y
+RUN rm -rf /debs
+
+COPY ["*.j2", "/usr/share/sonic/templates/"]
+COPY ["start.sh", "config.sh", "/usr/bin/"]
+COPY ["daemons", "/etc/quagga/"]
+
+ENTRYPOINT /usr/bin/config.sh   \
+        && /usr/bin/start.sh    \
+        && /bin/bash
```
Comment (stcheng) [inline dockers/docker-fpm-quagga/Dockerfile]: Remove this file. We could add Dockerfile to .gitignore if Dockerfile.j2 exists.

File: src/sonic-frr/Makefile [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+MAIN_TARGET = frr_$(FRR_VERSION)-dev_amd64.deb
+
+$(addprefix $(DEST)/, $(MAIN_TARGET)): $(DEST)/% :
+        # Cloning FRR repo if not already done
+	if [ -d "~/frr" ]; then \
+		git clone https://github.com/FRRouting/frr.git; \
+	fi
+
+	cp sonic_frr.rules frr/debian/rules
+	cp sonic_frr.install frr/debian/frr.install
```
Comment (stcheng) [inline src/sonic-frr/Makefile]: No need to check in these two files.

File: rules/frr.mk [hunk 4]
```diff
@@ -0,0 +1,9 @@
+# FRRouting (frr) package
+
+FRR_VERSION = 3.1
+export FRR_VERSION
+
+FRR = frr_$(FRR_VERSION)-dev_amd64.deb
```
Comment (lguohan) [inline rules/frr.mk]: >-dev [](start = 24, length = 4)

why the package name has dev? usually dev pacakge is for the library development files.

File: rules/docker-fpm-frr.mk [hunk 5]
```diff
@@ -0,0 +1,15 @@
+# docker image for fpm-frr
+
+DOCKER_FPM_FRR = docker-fpm-frr.gz
+$(DOCKER_FPM_FRR)_PATH = $(DOCKERS_PATH)/docker-fpm-frr
+$(DOCKER_FPM_FRR)_DEPENDS += $(FRR) $(SWSS)
+$(DOCKER_FPM_FRR)_LOAD_DOCKERS += $(DOCKER_CONFIG_ENGINE)
+SONIC_DOCKER_IMAGES += $(DOCKER_FPM_FRR)
```
Comment (lguohan) [inline rules/docker-fpm-frr.mk]: remove this as you have remove this in docker-fpm-gobgp.mk

File: rules/docker-fpm-frr.mk [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+$(DOCKER_FPM_FRR)_CONTAINER_NAME = bgp
+$(DOCKER_FPM_FRR)_RUN_OPT += --net=host --privileged -t
+$(DOCKER_FPM_FRR)_RUN_OPT += --volumes-from database
```
Comment (lguohan) [inline rules/docker-fpm-frr.mk]: this is no longer needed.

File: dockers/docker-fpm-frr/start.sh [hunk 7]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+
+service rsyslog start
```
Comment (lguohan) [inline dockers/docker-fpm-frr/start.sh]: suggest to add rm -f /var/run/rsyslogd.pid

File: dockers/docker-fpm-frr/bgpd.conf.j2 [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+hostname {{ inventory_hostname }}
+password zebra
+log syslog informational
+log facility local4
+! enable password {# {{ en_passwd }} TODO: param needed #}
+{% endblock system_init %}
+!
+{% block bgp_init %}
+!
+! bgp multiple-instance
```
Comment (lguohan) [inline dockers/docker-fpm-frr/bgpd.conf.j2]: should fpm-frr use same configuration as quagga or should it have a separate configuration?

Comment (rodnymolina) [inline dockers/docker-fpm-frr/bgpd.conf.j2]: Currently config-CLIs for Quagga and FRR are not identical (to say the least) -- there's an increasing feature gap. Overtime, as more features are added to FRR, this gap will become larger, so I would definitely vote for having different 'j2' templates for both routing-stacks.

File: dockers/docker-fpm-frr/start.sh (hunk 7 above)
Comment (rodnymolina) [inline dockers/docker-fpm-frr/start.sh]: done

File: rules/frr.mk (hunk 4 above)
Comment (rodnymolina) [inline rules/frr.mk]: I was originally pulling from dev/master branch, but i just changed to 'stable', so just got rid of 'dev' string.

File: dockers/docker-fpm-quagga/Dockerfile (hunk 2 above)
Comment (rodnymolina) [inline dockers/docker-fpm-quagga/Dockerfile]: done

File: dockers/docker-fpm-frr/Dockerfile (hunk 1 above)
Comment (rodnymolina) [inline dockers/docker-fpm-frr/Dockerfile]: done

File: dockers/docker-fpm-frr/bgpd.conf.j2 (hunk 8 above)
Comment (lguohan) [inline dockers/docker-fpm-frr/bgpd.conf.j2]: ok for me.

File: src/sonic-frr/Makefile [hunk 9]
```diff
@@ -0,0 +1,24 @@
+.ONESHELL:
+SHELL = /bin/bash
+.SHELLFLAGS += -e
+
+MAIN_TARGET = frr_$(FRR_VERSION)-dev_amd64.deb
+
+$(addprefix $(DEST)/, $(MAIN_TARGET)): $(DEST)/% :
+        # Cloning FRR repo if not already done
+	if [ -d "~/frr" ]; then \
+		git clone https://github.com/FRRouting/frr.git; \
```
Comment (lguohan) [inline src/sonic-frr/Makefile]: do you want to checkout a tag or this master branch, in this case the nightly build will be different as the frr moves forward.

File: src/sonic-frr/Makefile [hunk 10]
```diff
@@ -0,0 +1,24 @@
+.ONESHELL:
+SHELL = /bin/bash
```
Comment (lguohan) [inline src/sonic-frr/Makefile]: do we want to call it sonic-frr, so just frr?

File: src/sonic-frr/Makefile (hunk 9 above)
Comment (rodnymolina) [inline src/sonic-frr/Makefile]: In my latest patch (to be submitted soon), I'm now pulling frr's "stable/3.0" branch.

File: src/sonic-frr/Makefile (hunk 10 above)
Comment (rodnymolina) [inline src/sonic-frr/Makefile]: In my current proposal, 'sonic-frr' is the wrapper folder containing the SONiC specific information required to build frr. Whereas 'frr' subfolder contains the cloned frr repo. Please let me know if you have any other suggestion for the naming convention.

File: rules/docker-fpm-frr.mk (hunk 6 above)
Comment (lguohan) [inline rules/docker-fpm-frr.mk]: We do not need to get volumes-from database now.

File: rules/docker-fpm-frr.mk (hunk 5 above)
Comment (lguohan) [inline rules/docker-fpm-frr.mk]: in the rules/docker-fpm-gobgp.mk, you have removed "SONIC_DOCKER_IMAGES += $(DOCKER_FPM_GOBGP)", I am wondering why you still keep this in docker-fpm-frr.mk

File: src/sonic-frr/Makefile (hunk 10 above)
Comment (lguohan) [inline src/sonic-frr/Makefile]: fine by me.

File: rules/docker-fpm-frr.mk (hunk 6 above)
Comment (rodnymolina) [inline rules/docker-fpm-frr.mk]: understood, done

**Comment by lguohan** [comment]
Thanks for the PR.

**Comment by Ragsboss** [comment]
We are just users of SONiC but prefer FRR as the routing stack. What are our options?

# Build our own SONiC image with FRR routing stack - this is non-ideal because we want to use a stock image, not something we build on our own and who knows what differences our image will have.. plus we need to keep in sync with changes upstream etc - too costly
# Switch the routing stack at runtime - is this possible, supported and tested? Is there a documentation for this?
# Any other way?

Appreciate your thoughts.

**Comment by rodnymolina** [comment]
Right, AFAIK we are only building quagga-based images in the official sonic-jenkins page. The easiest way for you to go here is to build your own sonic-image with FRR routing-stack enabled.

Answering your second question: yes, you can switch from one routing-stack to the other at runtime, but be aware that this is not officially supported and that it will definitely have an impact on your service during the time you remove the old bgp container and you re-create the new one. From the host linux-shell in your sonic device do:

$ docker stop bgp

$ docker rm bgp

$ docker load < docker-fpm-frr.gz      [ you will find this docker-image in your build-server at sonic-buildimage/target/ folder ]

$ sudo sed -i 's/bgp docker-fpm-quagga/bgp docker-fpm-frr/g' /usr/bin/bgp.sh

$ sudo /usr/bin/bgp.sh start

That will recreate a new bgp container with frr as the routing-stack.

/Rodny


On Mar 28, 2018, at 12:50 PM, Ragsboss <notifications@github.com<mailto:notifications@github.com>> wrote:


We are just users of SONiC but prefer FRR as the routing stack. What are our options?

Build our own SONiC image with FRR routing stack - this is non-ideal because we want to use a stock image, not something we build on our own and who knows what differences our image will have.. plus we need to keep in sync with changes upstream etc - too costly
Switch the routing stack at runtime - is this possible, supported and tested? Is there a documentation for this?
Any other way?

Appreciate your thoughts.

**Comment by Ragsboss** [comment]
Thank you @rodnymolina. By runtime, I really meant as a one-time setup thing before the switch is put into service. I believe it's fair to say switching routing stacks causes disruption.

Its good to know there is a way to do this but can you elaborate on differences between choosing BGP stack at image build time vs. afterwards like you described above. I ask specifically because your note was cautionary - not sure if this is because FRR in general is not as tested as Quagga or due to some differences between binding FRR at image time vs. later.

**Pull request merged**